src/filters.ts

```diff
@@ -0,0 +1,253 @@
+/**
+ * Maps from the query to the corresponding type.
+ */
+type ChatMemberQueriesMap = {
+  admin: ChatMemberAdmin;
+  administrator: ChatMemberAdministrator;
+  creator: ChatMemberOwner;
+  free: ChatMemberFree;
+  in: ChatMemberIn;
+  out: ChatMemberOut;
+  regular: ChatMemberRegular;
+  kicked: ChatMemberBanned;
+  left: ChatMemberLeft;
+  member: ChatMemberMember;
+  restricted: ChatMemberRestricted;
+  restricted_in: ChatMemberRestrictedIn;
+  restricted_out: ChatMemberRestrictedOut;
+};
+
+type NormalizeChatMemberQueryCore<Q extends ChatMemberQuery> = (typeof chatMemberQueries)[Q][number];
+
+type MaybeArray<T> = T | T[];
+type NormalizeChatMemberQuery<
+  Q extends ChatMemberQuery,
+> = Q extends ChatMemberQuery ? NormalizeChatMemberQueryCore<Q>
+  : (Q extends ChatMemberQuery[] ? NormalizeChatMemberQuery<Q[number]>
+    : never);
+export type FilteredChatMember<
+  C extends ChatMember,
+  Q extends ChatMemberQuery,
+> = C & ChatMemberQueriesMap[NormalizeChatMemberQuery<Q>];
+
+/**
+ * Normalizes the query, returning the corresponding list of chat member
+ * statuses.
+ */
+function normalizeChatMemberQuery<T extends ChatMemberQuery>(
+  query: MaybeArray<T>,
+): NormalizeChatMemberQuery<T>[] {
+  if (Array.isArray(query)) {
+    const res = new Set<ChatMemberQuery>(
+      query.flatMap(normalizeChatMemberQuery),
+    );
+    return [...res] as NormalizeChatMemberQuery<T>[];
+  }
+
+  return [
+    ...chatMemberQueries[query],
+  ] as NormalizeChatMemberQuery<T>[];
+}
+
+/**
+ * Utility function to query the status of a chat member.
+ *
+ * Pass one of 'restricted_in', 'restricted_out', 'in', 'out', 'free', 'admin',
+ * 'regular', or one of the default Telegram statuses ('administrator',
+ * 'creator', 'kicked', 'left', 'member', 'restricted'), or an array of them.
+ *
+ * Returns true if the chat member matches the query.
+ */
+export function chatMemberIs<
+  C extends ChatMember,
+  Q extends ChatMemberQuery,
+>(
+  chatMember: C,
+  query: MaybeArray<Q>,
+): chatMember is FilteredChatMember<C, Q> {
+  const roles = normalizeChatMemberQuery(query);
+
+  if (chatMember.status === "restricted") {
+    if (roles.includes("restricted" as (typeof roles)[number])) {
+      return true;
+    } else if (chatMember.is_member) {
+      return roles.includes("restricted_in" as (typeof roles)[number]);
+    } else {
+      return roles.includes("restricted_out" as (typeof roles)[number]);
+    }
+  }
+
+  return roles.includes(chatMember.status as (typeof roles)[number]);
+}
+
+/**
+ * Filter context to only find updates of type 'my_chat_member' where the status
+ * transitions from oldStatus to newStatus.
+ *
+ * Example:
+ * ```typescript
+ * // listen for updates where the bot enters a group/supergroup
+ * bot.chatType(['group', 'supergroup']).filter(
+ *  myChatMemberFilter('out', 'in'),
+ *  (ctx) => {
+ *    const { old_chat_member: oldChatMember, new_chat_member: newChatMember } =
+ *      ctx.myChatMember;
+ *    // ...
+ *  },
+ * );
```

Comment:
This is poorly formatted. It is missing the end tag for the fenced markdown. Our website's CI system detected this: https://github.com/grammyjs/website/actions/runs/14077798929/job/39423885725?pr=1204

/cc @rojvv